File: Week-Wise-Assignments/Assignment 4/Compressed Solutions/C Programs/1_reverse_list.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* node = malloc(sizeof(struct Node));
    node->data = data;
    node->next = NULL;
    return node;
}

struct Node* reverse(struct Node* head) {
    struct Node* prev = NULL, *curr = head, *next;
    while (curr) {
        next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}

void display(struct Node* head) {
    while (head) {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}

int main() {
    struct Node* head = createNode(1);
    head->next = createNode(2);
    head->next->next = createNode(3);
    head->next->next->next = createNode(4);
    
    printf("Original: ");
    display(head);
    
    head = reverse(head);
    printf("Reversed: ");
    display(head);
    
    return 0;
}

File: Week-Wise-Assignments/Assignment 4/Compressed Solutions/Python Programs/1_reverse_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def reverse_iterative(head):
    prev, curr = None, head
    while curr:
        next_node = curr.next
        curr.next = prev
        prev, curr = curr, next_node
    return prev

def reverse_recursive(head):
    if not head or not head.next:
        return head
    new_head = reverse_recursive(head.next)
    head.next.next = head
    head.next = None
    return new_head

def display(head):
    while head:
        print(head.data, end=" -> " if head.next else " -> NULL\n")
        head = head.next

# Test
head = Node(1)
head.next = Node(2)
head.next.next = Node(3)
head.next.next.next = Node(4)

print("Original: ", end="")
display(head)

head = reverse_iterative(head)
print("Reversed: ", end="")
display(head)